src/document.c

```diff
@@ -930,36 +930,89 @@ typedef struct
 } FileData;
 
 
+static time_t get_mtime(const gchar *locale_filename)
+{
+	GError *error = NULL;
+	const gchar *err_msg = NULL;
+	time_t time = 0;
+
+	if (!file_prefs.use_safe_file_saving && file_prefs.use_gio_unsafe_file_saving)
+	{
+		GFile *file = g_file_new_for_path(locale_filename);
+		GFileInfo *info = g_file_query_info(file, G_FILE_ATTRIBUTE_TIME_MODIFIED, G_FILE_QUERY_INFO_NONE, NULL, &error);
+
+		if (info)
+		{
+			GTimeVal timeval;
+
+			g_file_info_get_modification_time(info, &timeval);
+			g_object_unref(info);
+			time = timeval.tv_sec;
+		}
+		else if (error)
+			err_msg = error->message;
+
+		g_object_unref(file);
+	}
+	else
+	{
+		struct stat st;
+
+		if (g_stat(locale_filename, &st) == 0)
+			time = st.st_mtime;
+		else
+			err_msg = g_strerror(errno);
+	}
+
+	if (err_msg)
+	{
+		gchar *utf8_filename = utils_get_utf8_from_locale(locale_filename);
+
+		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
+			utf8_filename, err_msg);
+		g_free(utf8_filename);
+	}
+
+	if (error)
+		g_error_free(error);
+
+	return time;
+}
+
+
 /* loads textfile data, verifies and converts to forced_enc or UTF-8. Also handles BOM. */
 static gboolean load_text_file(const gchar *locale_filename, const gchar *display_filename,
 	FileData *filedata, const gchar *forced_enc)
 {
 	GError *err = NULL;
-	struct stat st;
 
 	filedata->data = NULL;
 	filedata->len = 0;
 	filedata->enc = NULL;
 	filedata->bom = FALSE;
 	filedata->readonly = FALSE;
+	filedata->mtime = get_mtime(locale_filename);
 
-	if (g_stat(locale_filename, &st) != 0)
-	{
-		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
-			display_filename, g_strerror(errno));
+	if (filedata->mtime == 0)
```

Comment:
I might be paranoid, but what if the mtime is 1970-01-01 00:00:00?  I could even imagine a FS without mtime support always reporting 0 here or some other weird behavior, and this would be a real problem as it aborts loading.